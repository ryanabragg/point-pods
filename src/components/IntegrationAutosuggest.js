import React from 'react';
import PropTypes from 'prop-types';
import Autosuggest from 'react-autosuggest';
import match from 'autosuggest-highlight/match';
import parse from 'autosuggest-highlight/parse';
import { withStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import MenuItem from '@material-ui/core/MenuItem';
import TextField from '@material-ui/core/TextField';
import IconButton from '@material-ui/core/IconButton';

const styles = theme => ({
  root: {
    display: 'flex',
    flexGrow: 0,
    flexWrap: 'nowrap',
    alignItems: 'center',
  },
  container: {
    flexGrow: 1,
    position: 'relative',
  },
  suggestionsContainerOpen: {
    position: 'absolute',
    zIndex: 1,
    marginTop: theme.spacing.unit,
    left: 0,
    right: 0,
  },
  suggestion: {
    display: 'block',
  },
  suggestionsList: {
    margin: 0,
    padding: 0,
    listStyleType: 'none',
  },
});

class IntegrationAutosuggest extends React.Component {
  state = {
    value: '',
    suggestions: [],
  };

  handleChange = (event, { newValue }) => {
    this.setState({ value: newValue });
  };

  getSuggestionsList = (value) => {
    const regex = RegExp('\\b' + value.toLowerCase());
    let count = 0;
    if(value.length === 0)
      return [];
    return this.props.items.filter(item => {
      const match = count < this.props.listLength && regex.test(item.label.toLowerCase());
      if(match)
        count++;
      return match;
    });
  };

  handleSuggestionSelected = (event, { suggestion, suggestionValue, suggestionIndex, sectionIndex, method }) => {
    this.props.onSelect(suggestion);
    if(this.props.clearOnSelect)
      this.setState({ value: '', suggestions: [] });
  };

  handleSuggestionButton = (event) => {
    this.props.onSelect({id: null, label: this.state.value});
    if(this.props.clearOnSelect)
      this.setState({ value: '', suggestions: [] });
  };

  suggestionValue = (suggestion) => suggestion.label;

  handleSuggestionsFetchRequested = ({ value }) => {
    this.setState({
      suggestions: this.getSuggestionsList(value),
    });
  };

  handleSuggestionsClearRequested = () => {
    this.setState({
      suggestions: this.props.items,
    });
  };

  renderInputComponent = (props) => {
    const { classes, ref, ...other } = props;
    return (
      <TextField
        fullWidth
        InputProps={{
          inputRef: ref,
          classes: {
            input: classes.input,
          },
          ...other,
        }}
      />
    );
  };

  renderSuggestionsContainer = (props) => {
    const { containerProps, children } = props;
    return (
      <Paper {...containerProps} square>
        {children}
      </Paper>
    );
  };

  renderSuggestion = (suggestion, { query, isHighlighted }) => {
    const matches = match(suggestion.label, query);
    const parts = parse(suggestion.label, matches);
    return (
      <MenuItem selected={isHighlighted} component="div">
        {parts.map((part, index) => {
          return part.highlight ? (
            <span key={String(index)} style={{ fontWeight: 500 }}>
              {part.text}
            </span>
          ) : (
            <strong key={String(index)} style={{ fontWeight: 300 }}>
              {part.text}
            </strong>
          );
        })}
      </MenuItem>
    );
  };

  render() {
    const { classes, buttonIcon } = this.props;

    return (
      <div className={classes.root}>
        <Autosuggest
          theme={{
            container: classes.container,
            suggestionsContainerOpen: classes.suggestionsContainerOpen,
            suggestionsList: classes.suggestionsList,
            suggestion: classes.suggestion,
          }}
          renderInputComponent={this.renderInputComponent}
          inputProps={{
            classes,
            placeholder: 'Add a player',
            value: this.state.value,
            onChange: this.handleChange,
          }}
          suggestions={this.state.suggestions}
          getSuggestionValue={this.suggestionValue}
          onSuggestionsFetchRequested={this.handleSuggestionsFetchRequested}
          onSuggestionsClearRequested={this.handleSuggestionsClearRequested}
          onSuggestionSelected={this.handleSuggestionSelected}
          renderSuggestionsContainer={this.renderSuggestionsContainer}
          renderSuggestion={this.renderSuggestion}
        />
        {buttonIcon &&
          <IconButton color='inherit' onClick={this.handleSuggestionButton}>
            {buttonIcon}
          </IconButton>
        }
      </div>
    );
  }
}

IntegrationAutosuggest.defaultProps = {
  items: [],
  listLength: 5,
  onSelect: ({id, label}) => null,
  clearOnSelect: false,
  buttonIcon: null,
}

IntegrationAutosuggest.propTypes = {
  classes: PropTypes.object.isRequired,
  items: PropTypes.array,
  listLength: PropTypes.number,
  onSelect: PropTypes.func,
  clearOnSelect: PropTypes.bool,
  buttonIcon: PropTypes.element,
}

export default withStyles(styles, { withTheme: true })(IntegrationAutosuggest);
