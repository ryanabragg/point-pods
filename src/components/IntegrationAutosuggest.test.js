import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import IntegrationAutosuggest from './IntegrationAutosuggest';
import Autosuggest from 'react-autosuggest';
import IconButton from '@material-ui/core/IconButton';

const suggestions = [
  'Victoria',
  'Alexander',
  'Zara',
  'Kara',
  'Katka',
  'Alexandra',
  'Kat',
  'Kaori',
  'Kaya',
  'Jaya',
  'Kyrie',
  'Krystal',
].map(s => ({ id: s, label: s }));

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<IntegrationAutosuggest />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test('matches the prior snapshot', (done) => {
  const component = renderer.create(<IntegrationAutosuggest />);
  expect(component.toJSON()).toMatchSnapshot();
  component.update(<IntegrationAutosuggest items={suggestions} />);
  setTimeout(() => {
    try {
      expect(component.toJSON()).toMatchSnapshot();
      done();
    } catch (error) {
      done.fail(error);
    }
  }, 100);
});

describe('integration', () => {
  test('integration props', () => {
    const component = shallow(<IntegrationAutosuggest />);
    let unHOC = component.dive();
    expect(unHOC.find(Autosuggest).prop('renderInputComponent')).toEqual(unHOC.instance().renderInputComponent);
    expect(unHOC.find(Autosuggest).prop('inputProps').placeholder).toEqual('Add a player');
    expect(unHOC.find(Autosuggest).prop('inputProps').value).toEqual(unHOC.state('value'));
    expect(unHOC.find(Autosuggest).prop('inputProps').onChange).toEqual(unHOC.instance().handleChange);
    expect(unHOC.find(Autosuggest).prop('suggestions')).toEqual(unHOC.state('suggestions'));
    expect(unHOC.find(Autosuggest).prop('getSuggestionValue')).toEqual(unHOC.instance().suggestionValue);
    expect(unHOC.find(Autosuggest).prop('onSuggestionsFetchRequested')).toEqual(unHOC.instance().handleSuggestionsFetchRequested);
    expect(unHOC.find(Autosuggest).prop('onSuggestionsClearRequested')).toEqual(unHOC.instance().handleSuggestionsClearRequested);
    expect(unHOC.find(Autosuggest).prop('onSuggestionSelected')).toEqual(unHOC.instance().handleSuggestionSelected);
    expect(unHOC.find(Autosuggest).prop('renderSuggestionsContainer')).toEqual(unHOC.instance().renderSuggestionsContainer);
    expect(unHOC.find(Autosuggest).prop('renderSuggestion')).toEqual(unHOC.instance().renderSuggestion);
  });

  test('add button', () => {
    const component = shallow(<IntegrationAutosuggest items={suggestions} />);
    let unHOC = component.dive();
    expect(unHOC.find(IconButton).length).toBe(0);
    component.setProps({ buttonIcon: 'test' });
    unHOC = component.dive();
    expect(unHOC.find(IconButton).prop('onClick')).toEqual(unHOC.instance().handleSuggestionButton);
  });
});

describe('actions', () => {
  test('select a suggestion', () => {
    const selection = { id: 'testID', label: 'test' };
    const onSelect = jest.fn(r => r);
    const component = shallow(<IntegrationAutosuggest onSelect={onSelect} />);
    let unHOC = component.dive();
    unHOC.instance().handleSuggestionSelected({event: 'test'}, { suggestion: selection });
    expect(onSelect.mock.calls[0][0]).toEqual(selection);
  });

  test('add a new value', () => {
    const selection = { id: null, label: 'test' };
    const onSelect = jest.fn(r => r);
    const component = shallow(<IntegrationAutosuggest onSelect={onSelect} />);
    let unHOC = component.dive();
    unHOC.setState({
      value: selection.label,
    });
    unHOC.instance().handleSuggestionButton();
    expect(unHOC.state('value')).toBe(selection.label);
    expect(onSelect.mock.calls[0][0]).toEqual(selection);
    unHOC.setProps({
      clearOnSelect: true,
    });
    unHOC.instance().handleSuggestionButton();
    expect(unHOC.state('value')).toBe('');
    expect(onSelect.mock.calls[0][0]).toEqual(selection);
  });
});
