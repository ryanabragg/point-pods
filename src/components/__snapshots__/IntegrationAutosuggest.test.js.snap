// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the prior snapshot 1`] = `
<div
  className="IntegrationAutosuggest-root-50"
>
  <div
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-owns="react-autowhatever-1"
    className="IntegrationAutosuggest-container-51"
    role="combobox"
  >
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-55 MuiFormControl-fullWidth-58"
    >
      <div
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-controls="react-autowhatever-1"
        className="MuiInput-root-59 MuiInput-fullWidth-66 MuiInput-formControl-60 MuiInput-underline-63"
        style={Object {}}
      >
        <input
          aria-invalid={false}
          autoComplete="off"
          autoFocus={undefined}
          className="MuiInput-input-67"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={undefined}
          placeholder="Add a player"
          readOnly={undefined}
          required={undefined}
          rows={undefined}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="MuiPaper-root-72 MuiPaper-elevation2-76"
      id="react-autowhatever-1"
      role="listbox"
      style={Object {}}
    />
  </div>
</div>
`;

exports[`matches the prior snapshot 2`] = `
<div
  className="IntegrationAutosuggest-root-50"
>
  <div
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-owns="react-autowhatever-1"
    className="IntegrationAutosuggest-container-51"
    role="combobox"
  >
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-55 MuiFormControl-fullWidth-58"
    >
      <div
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-controls="react-autowhatever-1"
        className="MuiInput-root-59 MuiInput-fullWidth-66 MuiInput-formControl-60 MuiInput-underline-63"
        style={Object {}}
      >
        <input
          aria-invalid={false}
          autoComplete="off"
          autoFocus={undefined}
          className="MuiInput-input-67"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={undefined}
          placeholder="Add a player"
          readOnly={undefined}
          required={undefined}
          rows={undefined}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="MuiPaper-root-72 MuiPaper-elevation2-76"
      id="react-autowhatever-1"
      role="listbox"
      style={Object {}}
    />
  </div>
</div>
`;
